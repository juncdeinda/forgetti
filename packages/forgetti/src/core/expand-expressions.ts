import * as t from '@babel/types';
import type * as babel from '@babel/core';
import type { ComponentNode, StateContext } from './types';
import { getHookCallType } from './get-hook-call-type';
import { isPathValid } from './checks';
import unwrapNode from './unwrap-node';

function isStatementValid(path: babel.NodePath): boolean {
  if (path) {
    switch (path.node.type) {
      case 'ForInStatement':
      case 'ForOfStatement':
      case 'ForStatement':
      case 'WhileStatement':
      case 'DoWhileStatement':
        return false;
      default:
        return true;
    }
  }
  return false;
}

function isInValidExpression(path: babel.NodePath): boolean {
  let current = path.parentPath;
  let prev = path;
  while (current) {
    if (
      isPathValid(current, t.isConditionalExpression) &&
      (current.get('consequent').node === prev.node ||
        current.get('alternate').node === prev.node)
    ) {
      return false;
    }
    if (
      isPathValid(current, t.isLogicalExpression) &&
      current.get('right').node === prev.node
    ) {
      return false;
    }
    prev = current;
    current = current.parentPath;
  }
  return true;
}

const UNDEFINED_LITERAL = t.unaryExpression('void', t.numericLiteral(0));

function transformOptionalCall(
  path: babel.NodePath<t.OptionalCallExpression>,
): t.Expression {
  const unwrappedID = unwrapNode(path.node.callee, t.isIdentifier);
  if (unwrappedID) {
    return t.conditionalExpression(
      t.binaryExpression('==', unwrappedID, t.nullLiteral()),
      UNDEFINED_LITERAL,
      t.callExpression(unwrappedID, path.node.arguments),
    );
  }
  const temp = path.scope.generateUidIdentifier('nullish');
  path.scope.push({
    kind: 'let',
    id: temp,
  });
  const unwrappedCallee =
    unwrapNode(path.node.callee, t.isMemberExpression) ||
    unwrapNode(path.node.callee, t.isOptionalMemberExpression);
  if (unwrappedCallee) {
    let unwrappedObject = unwrapNode(unwrappedCallee.object, t.isIdentifier);
    if (!unwrappedObject) {
      unwrappedObject = path.scope.generateUidIdentifier('object');
      path.scope.push({
        kind: 'let',
        id: unwrappedObject,
      });
      unwrappedCallee.object = t.assignmentExpression(
        '=',
        unwrappedObject,
        unwrappedCallee.object,
      );
    }
    return t.conditionalExpression(
      t.binaryExpression(
        '==',
        t.assignmentExpression('=', temp, unwrappedCallee),
        t.nullLiteral(),
      ),
      UNDEFINED_LITERAL,
      t.callExpression(t.memberExpression(temp, t.identifier('call')), [
        unwrappedObject,
        ...path.node.arguments,
      ]),
    );
  }
  return t.conditionalExpression(
    t.binaryExpression(
      '==',
      t.assignmentExpression('=', temp, path.node.callee),
      t.nullLiteral(),
    ),
    UNDEFINED_LITERAL,
    t.callExpression(temp, path.node.arguments),
  );
}

function transformOptionalMember(
  path: babel.NodePath<t.OptionalMemberExpression>,
): t.Expression {
  const unwrappedID = unwrapNode(path.node.object, t.isIdentifier);
  if (unwrappedID) {
    return t.conditionalExpression(
      t.binaryExpression('==', unwrappedID, t.nullLiteral()),
      UNDEFINED_LITERAL,
      t.memberExpression(unwrappedID, path.node.property, path.node.computed),
    );
  }
  const temp = path.scope.generateUidIdentifier('nullish');
  path.scope.push({
    kind: 'let',
    id: temp,
  });
  return t.conditionalExpression(
    t.binaryExpression(
      '==',
      t.assignmentExpression('=', temp, path.node.object),
      t.nullLiteral(),
    ),
    UNDEFINED_LITERAL,
    t.memberExpression(temp, path.node.property, path.node.computed),
  );
}

export function expandExpressions(
  ctx: StateContext,
  path: babel.NodePath<ComponentNode>,
): void {
  if (
    path.node.type === 'ArrowFunctionExpression' &&
    path.node.body.type !== 'BlockStatement'
  ) {
    path.node.body = t.blockStatement([t.returnStatement(path.node.body)]);
  }
  path.traverse({
    OptionalCallExpression(p) {
      const parent = p.getFunctionParent();
      const statement = p.getStatementParent();

      if (parent === path && statement) {
        p.replaceWith(transformOptionalCall(p));
      }
    },
    OptionalMemberExpression(p) {
      const parent = p.getFunctionParent();
      const statement = p.getStatementParent();

      if (parent === path && statement) {
        p.replaceWith(transformOptionalMember(p));
      }
    },
    AssignmentExpression(p) {
      const parent = p.getFunctionParent();
      const statement = p.getStatementParent();

      if (
        parent === path &&
        statement &&
        isStatementValid(statement) &&
        isInValidExpression(p) &&
        !isPathValid(p.parentPath, t.isStatement) &&
        !isPathValid(p.parentPath, t.isVariableDeclarator)
      ) {
        const id = p.scope.generateUidIdentifier('hoisted');
        statement.insertBefore(
          t.variableDeclaration('let', [t.variableDeclarator(id, p.node)]),
        );
        p.replaceWith(id);
      }
    },
    CallExpression(p) {
      const parent = p.getFunctionParent();
      const statement = p.getStatementParent();

      if (
        parent === path &&
        statement &&
        !isPathValid(p.parentPath, t.isStatement) &&
        !isPathValid(p.parentPath, t.isVariableDeclarator)
      ) {
        const hookType = getHookCallType(ctx, p);
        if (hookType === 'custom' || hookType === 'effect') {
          const id = p.scope.generateUidIdentifier('hoisted');
          statement.insertBefore(
            t.variableDeclaration('let', [t.variableDeclarator(id, p.node)]),
          );
          p.replaceWith(id);
        }
      }
    },
  });

  path.scope.crawl();
}
