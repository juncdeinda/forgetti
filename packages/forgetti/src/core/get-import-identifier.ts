import type * as babel from '@babel/core';
import type * as t from '@babel/types';
import { addDefault, addNamed } from '@babel/helper-module-imports';
import type { ImportDefinition } from './presets';
import type { StateContext } from './types';

export default function getImportIdentifier(
  ctx: StateContext,
  path: babel.NodePath,
  definition: ImportDefinition,
): t.Identifier {
  const name = definition.kind === 'default' ? 'default' : definition.name;
  const target = `${definition.source}[${name}]`;
  const current = ctx.imports.get(target);
  if (current) {
    return current;
  }
  const newID =
    definition.kind === 'named'
      ? addNamed(path, definition.name, definition.source)
      : addDefault(path, definition.source);
  ctx.imports.set(target, newID);
  return newID;
}
