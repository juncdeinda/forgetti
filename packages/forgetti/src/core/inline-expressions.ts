import type * as babel from '@babel/core';
import * as t from '@babel/types';
import type { ComponentNode } from './types';
import { isPathValid } from './checks';

function isInValidExpression(path: babel.NodePath): boolean {
  let current = path.parentPath;
  let prev = path;
  while (current) {
    if (
      isPathValid(current, t.isConditionalExpression) &&
      (current.get('consequent').node === prev.node ||
        current.get('alternate').node === prev.node)
    ) {
      return false;
    }
    if (
      isPathValid(current, t.isLogicalExpression) &&
      current.get('right').node === prev.node
    ) {
      return false;
    }
    prev = current;
    current = current.parentPath;
  }
  return true;
}

function inlineExpression(
  parentPath: babel.NodePath<ComponentNode>,
  path: babel.NodePath<t.Expression>,
) {
  if (
    path.getFunctionParent() === parentPath &&
    isPathValid(path, t.isIdentifier)
  ) {
    const binding = path.scope.getBinding(path.node.name);
    if (binding?.referenced && binding.referencePaths.length === 1) {
      switch (binding.kind) {
        case 'const':
        case 'let':
        case 'var': {
          // move the node to the reference
          const ref = binding.referencePaths[0];
          if (
            isInValidExpression(ref) &&
            isPathValid(binding.path, t.isVariableDeclarator) &&
            binding.path.node.init &&
            isPathValid(binding.path.get('id'), t.isIdentifier) &&
            binding.path.scope.getBlockParent() === ref.scope.getBlockParent()
          ) {
            ref.replaceWith(binding.path.node.init);
            binding.path.remove();
          }
          break;
        }
      }
    }
  }
}

export function inlineExpressions(path: babel.NodePath<ComponentNode>): void {
  path.traverse({
    Expression(p) {
      inlineExpression(path, p);
    },
  });
  path.scope.crawl();
}
