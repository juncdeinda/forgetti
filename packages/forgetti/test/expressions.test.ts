import * as babel from '@babel/core';
import { describe, it } from 'vitest';
import type { Options } from '../src';
import plugin from '../src';

const options: Options = {
  preset: 'react',
};

async function compile(code: string): Promise<string> {
  const result = await babel.transformAsync(code, {
    plugins: [[plugin, options]],
    parserOpts: {
      plugins: ['jsx'],
    },
  });

  return result?.code ?? '';
}

describe('expressions', () => {
  it('should optimize guaranteed literals', async ({ expect }) => {
    const code = `
function Example(props) {
  return 1 + 2;
}
`;
    expect(await compile(code)).toMatchSnapshot();
  });
  it('should optimize identifiers', async ({ expect }) => {
    const code = `
function Example(props) {
  return props;
}
`;
    expect(await compile(code)).toMatchSnapshot();
  });
  it('should optimize member expressions', async ({ expect }) => {
    const code = `
function Example(props) {
  return props.example;
}
`;
    expect(await compile(code)).toMatchSnapshot();
  });
  it('should optimize conditional expressions', async ({ expect }) => {
    const code = `
function Example(props) {
  return props.a ? props.b : props.c;
}
`;
    expect(await compile(code)).toMatchSnapshot();
  });
  it('should optimize binary expressions', async ({ expect }) => {
    const code = `
function Example(props) {
  return props.a + props.b;
}
`;
    expect(await compile(code)).toMatchSnapshot();
  });
  it('should optimize logical expressions', async ({ expect }) => {
    const code = `
function Example(props) {
  return props.a && props.b;
}
`;
    expect(await compile(code)).toMatchSnapshot();
  });
  it('should optimize unary expressions', async ({ expect }) => {
    const code = `
function Example(props) {
  return !props.a;
}
`;
    expect(await compile(code)).toMatchSnapshot();
  });
  it('should optimize call expressions', async ({ expect }) => {
    const code = `
function Example(props) {
  return props.call();
}
`;
    expect(await compile(code)).toMatchSnapshot();
  });
  it('should optimize function expressions', async ({ expect }) => {
    const code = `
function Example(props) {
  const callback = () => {
    console.log(props.message);
  };
}
`;
    expect(await compile(code)).toMatchSnapshot();
  });
  it('should optimize assignment expressions', async ({ expect }) => {
    const code = `
function Example(props) {
  let a, b, c;

  a = b = c = props.x;
}
`;
    expect(await compile(code)).toMatchSnapshot();
  });
  it('should optimize array expressions', async ({ expect }) => {
    const code = `
function Example(props) {
  return [props.a, props.b, ...props.c];
}
`;
    expect(await compile(code)).toMatchSnapshot();
  });
  it('should optimize object expressions', async ({ expect }) => {
    const code = `
function Example(props) {
  return { a: props.a, b: props.b, ...props.c };
}
`;
    expect(await compile(code)).toMatchSnapshot();
  });
  it('should optimize new expressions', async ({ expect }) => {
    const code = `
function Example(props) {
  return new X(props);
}
`;
    expect(await compile(code)).toMatchSnapshot();
  });
  it('should optimize sequence expressions', async ({ expect }) => {
    const code = `
function Example(props) {
  return props.a(), props.b();
}
`;
    expect(await compile(code)).toMatchSnapshot();
  });
  it('should optimize template literals', async ({ expect }) => {
    const code = `
function Example(props) {
  return \`\${props.a()}, \${props.b()}\`;
}
`;
    expect(await compile(code)).toMatchSnapshot();
  });
  it('should optimize tagged templates', async ({ expect }) => {
    const code = `
function Example(props) {
  return props.tag\`\${props.a()}, \${props.b()}\`;
}
`;
    expect(await compile(code)).toMatchSnapshot();
  });
  it('should optimize JSX Element', async ({ expect }) => {
    const code = `
function Example(props) {
  return (
    <div>
      <h1 title={props.title}>Title: {props.title}</h1>
      {props.children}
    </div>
  );
}
`;
    expect(await compile(code)).toMatchSnapshot();
  });
  it('should optimize JSX Fragment', async ({ expect }) => {
    const code = `
function Example(props) {
  return (
    <>
      <h1 title={props.title}>Title: {props.title}</h1>
      {props.children}
    </>
  );
}
`;
    expect(await compile(code)).toMatchSnapshot();
  });
  it('should optimize optional member expression', async ({ expect }) => {
    const code = `
function Example(props) {
  return props?.a?.b?.c;
}
`;
    expect(await compile(code)).toMatchSnapshot();
  });
  it('should optimize optional call expression', async ({ expect }) => {
    const code = `
function Example(props) {
  return props.a?.(props.b, props.c);
}
`;
    expect(await compile(code)).toMatchSnapshot();
  });
});
