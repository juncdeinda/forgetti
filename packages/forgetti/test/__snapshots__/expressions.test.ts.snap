// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`expressions > should optimize JSX Element 1`] = `
"import { useMemo as _useMemo } from \\"react\\";
import { $$cache as _$$cache } from \\"forgetti/runtime\\";
import { $$equals as _$$equals } from \\"forgetti/runtime\\";
import { memo as _memo } from \\"react\\";
import { $$memo as _$$memo } from \\"forgetti/runtime\\";
const _Example = _$$memo(_memo, \\"_Example\\", _values => <div>
      <h1 title={_values[0]}>Title: {_values[1]}</h1>
      {_values[2]}
    </div>);
function Example(props) {
  let _cache = _$$cache(_useMemo, 10),
    _equals = _$$equals(_cache, 0, props),
    _value = _equals ? _cache[0] : _cache[0] = props,
    _value2 = _equals ? _cache[1] : _cache[1] = _value.title,
    _equals2 = _$$equals(_cache, 2, _value2),
    _value3 = _equals2 ? _cache[2] : _cache[2] = _value2,
    _value4 = _equals ? _cache[3] : _cache[3] = _value.title,
    _equals3 = _$$equals(_cache, 4, _value4),
    _value5 = _equals3 ? _cache[4] : _cache[4] = _value4,
    _value6 = _equals ? _cache[5] : _cache[5] = _value.children,
    _equals4 = _$$equals(_cache, 6, _value6),
    _value7 = _equals4 ? _cache[6] : _cache[6] = _value6,
    _value8 = _equals2 && _equals3 && _equals4 ? _cache[7] : _cache[7] = [_value3, _value5, _value7],
    _equals6 = _$$equals(_cache, 8, _value8),
    _value9 = _equals6 ? _cache[8] : _cache[8] = _value8;
  return /*@forgetti jsx*/_equals6 ? _cache[9] : _cache[9] = <_Example v={_value9} />;
}"
`;

exports[`expressions > should optimize JSX Fragment 1`] = `
"import { useMemo as _useMemo } from \\"react\\";
import { $$cache as _$$cache } from \\"forgetti/runtime\\";
import { $$equals as _$$equals } from \\"forgetti/runtime\\";
import { memo as _memo } from \\"react\\";
import { $$memo as _$$memo } from \\"forgetti/runtime\\";
const _Example = _$$memo(_memo, \\"_Example\\", _values => <>
      <h1 title={_values[0]}>Title: {_values[1]}</h1>
      {_values[2]}
    </>);
function Example(props) {
  let _cache = _$$cache(_useMemo, 10),
    _equals = _$$equals(_cache, 0, props),
    _value = _equals ? _cache[0] : _cache[0] = props,
    _value2 = _equals ? _cache[1] : _cache[1] = _value.title,
    _equals2 = _$$equals(_cache, 2, _value2),
    _value3 = _equals2 ? _cache[2] : _cache[2] = _value2,
    _value4 = _equals ? _cache[3] : _cache[3] = _value.title,
    _equals3 = _$$equals(_cache, 4, _value4),
    _value5 = _equals3 ? _cache[4] : _cache[4] = _value4,
    _value6 = _equals ? _cache[5] : _cache[5] = _value.children,
    _equals4 = _$$equals(_cache, 6, _value6),
    _value7 = _equals4 ? _cache[6] : _cache[6] = _value6,
    _value8 = _equals2 && _equals3 && _equals4 ? _cache[7] : _cache[7] = [_value3, _value5, _value7],
    _equals6 = _$$equals(_cache, 8, _value8),
    _value9 = _equals6 ? _cache[8] : _cache[8] = _value8;
  return /*@forgetti jsx*/_equals6 ? _cache[9] : _cache[9] = <_Example v={_value9} />;
}"
`;

exports[`expressions > should optimize array expressions 1`] = `
"import { useMemo as _useMemo } from \\"react\\";
import { $$cache as _$$cache } from \\"forgetti/runtime\\";
import { $$equals as _$$equals } from \\"forgetti/runtime\\";
function Example(props) {
  let _cache = _$$cache(_useMemo, 8),
    _equals = _$$equals(_cache, 0, props),
    _value = _equals ? _cache[0] : _cache[0] = props,
    _value2 = _equals ? _cache[1] : _cache[1] = _value.a,
    _equals2 = _$$equals(_cache, 2, _value2),
    _value3 = _equals2 ? _cache[2] : _cache[2] = _value2,
    _value4 = _equals ? _cache[3] : _cache[3] = _value.b,
    _equals3 = _$$equals(_cache, 4, _value4),
    _value5 = _equals3 ? _cache[4] : _cache[4] = _value4,
    _value6 = _equals ? _cache[5] : _cache[5] = _value.c,
    _equals4 = _$$equals(_cache, 6, _value6),
    _value7 = _equals4 ? _cache[6] : _cache[6] = _value6;
  return _equals2 && _equals3 && _equals4 ? _cache[7] : _cache[7] = [_value3, _value5, ..._value7];
}"
`;

exports[`expressions > should optimize assignment expressions 1`] = `
"import { useMemo as _useMemo } from \\"react\\";
import { $$cache as _$$cache } from \\"forgetti/runtime\\";
import { $$equals as _$$equals } from \\"forgetti/runtime\\";
function Example(props) {
  let _cache = _$$cache(_useMemo, 5),
    a,
    b,
    c,
    _equals = _$$equals(_cache, 0, props),
    _value = _equals ? _cache[0] : _cache[0] = props,
    _value2 = _equals ? _cache[1] : _cache[1] = _value.x,
    _hoisted2 = c = _$$equals(_cache, 2, _value2) ? _cache[2] : _cache[2] = _value2,
    _hoisted = b = _$$equals(_cache, 3, _hoisted2) ? _cache[3] : _cache[3] = _hoisted2;
  a = _$$equals(_cache, 4, _hoisted) ? _cache[4] : _cache[4] = _hoisted;
}"
`;

exports[`expressions > should optimize binary expressions 1`] = `
"import { useMemo as _useMemo } from \\"react\\";
import { $$cache as _$$cache } from \\"forgetti/runtime\\";
import { $$equals as _$$equals } from \\"forgetti/runtime\\";
function Example(props) {
  let _cache = _$$cache(_useMemo, 6),
    _equals = _$$equals(_cache, 0, props),
    _value = _equals ? _cache[0] : _cache[0] = props,
    _value2 = _equals ? _cache[1] : _cache[1] = _value.a,
    _equals2 = _$$equals(_cache, 2, _value2),
    _value3 = _equals2 ? _cache[2] : _cache[2] = _value2,
    _value4 = _equals ? _cache[3] : _cache[3] = _value.b,
    _equals3 = _$$equals(_cache, 4, _value4),
    _value5 = _equals3 ? _cache[4] : _cache[4] = _value4;
  return _equals2 && _equals3 ? _cache[5] : _cache[5] = _value3 + _value5;
}"
`;

exports[`expressions > should optimize call expressions 1`] = `
"import { useMemo as _useMemo } from \\"react\\";
import { $$cache as _$$cache } from \\"forgetti/runtime\\";
import { $$equals as _$$equals } from \\"forgetti/runtime\\";
function Example(props) {
  let _cache = _$$cache(_useMemo, 2),
    _equals = _$$equals(_cache, 0, props),
    _value = _equals ? _cache[0] : _cache[0] = props;
  return _equals ? _cache[1] : _cache[1] = _value.call();
}"
`;

exports[`expressions > should optimize conditional expressions 1`] = `
"import { useMemo as _useMemo } from \\"react\\";
import { $$cache as _$$cache } from \\"forgetti/runtime\\";
import { $$branch as _$$branch } from \\"forgetti/runtime\\";
import { $$equals as _$$equals } from \\"forgetti/runtime\\";
function Example(props) {
  let _cache = _$$cache(_useMemo, 4),
    _equals = _$$equals(_cache, 0, props),
    _value2 = _equals ? _cache[0] : _cache[0] = props,
    _value;
  if (_equals ? _cache[1] : _cache[1] = _value2.a) {
    let _cache2 = _$$branch(_cache, 2, 2),
      _equals2 = _$$equals(_cache2, 0, props),
      _value4 = _equals2 ? _cache2[0] : _cache2[0] = props;
    _value = _equals2 ? _cache2[1] : _cache2[1] = _value4.b;
  } else {
    let _cache3 = _$$branch(_cache, 3, 2),
      _equals3 = _$$equals(_cache3, 0, props),
      _value6 = _equals3 ? _cache3[0] : _cache3[0] = props;
    _value = _equals3 ? _cache3[1] : _cache3[1] = _value6.c;
  }
  return _value;
}"
`;

exports[`expressions > should optimize function expressions 1`] = `
"import { useMemo as _useMemo } from \\"react\\";
import { $$cache as _$$cache } from \\"forgetti/runtime\\";
import { $$equals as _$$equals } from \\"forgetti/runtime\\";
function Example(props) {
  let _cache = _$$cache(_useMemo, 2),
    _equals = _$$equals(_cache, 0, props),
    _value = _equals ? _cache[0] : _cache[0] = props;
  const callback = _equals ? _cache[1] : _cache[1] = () => {
    console.log(props.message);
  };
}"
`;

exports[`expressions > should optimize guaranteed literals 1`] = `
"import { useMemo as _useMemo } from \\"react\\";
import { $$cache as _$$cache } from \\"forgetti/runtime\\";
function Example(props) {
  let _cache = _$$cache(_useMemo, 1);
  return 0 in _cache ? _cache[0] : _cache[0] = 1 + 2;
}"
`;

exports[`expressions > should optimize identifiers 1`] = `
"import { useMemo as _useMemo } from \\"react\\";
import { $$cache as _$$cache } from \\"forgetti/runtime\\";
import { $$equals as _$$equals } from \\"forgetti/runtime\\";
function Example(props) {
  let _cache = _$$cache(_useMemo, 1);
  return _$$equals(_cache, 0, props) ? _cache[0] : _cache[0] = props;
}"
`;

exports[`expressions > should optimize logical expressions 1`] = `
"import { useMemo as _useMemo } from \\"react\\";
import { $$cache as _$$cache } from \\"forgetti/runtime\\";
import { $$branch as _$$branch } from \\"forgetti/runtime\\";
import { $$equals as _$$equals } from \\"forgetti/runtime\\";
function Example(props) {
  let _cache = _$$cache(_useMemo, 3),
    _equals = _$$equals(_cache, 0, props),
    _value = _equals ? _cache[0] : _cache[0] = props,
    _condition = _equals ? _cache[1] : _cache[1] = _value.a;
  if (_condition) {
    let _cache2 = _$$branch(_cache, 2, 2),
      _equals2 = _$$equals(_cache2, 0, props),
      _value3 = _equals2 ? _cache2[0] : _cache2[0] = props;
    _condition = _equals2 ? _cache2[1] : _cache2[1] = _value3.b;
  }
  return _condition;
}"
`;

exports[`expressions > should optimize member expressions 1`] = `
"import { useMemo as _useMemo } from \\"react\\";
import { $$cache as _$$cache } from \\"forgetti/runtime\\";
import { $$equals as _$$equals } from \\"forgetti/runtime\\";
function Example(props) {
  let _cache = _$$cache(_useMemo, 2),
    _equals = _$$equals(_cache, 0, props),
    _value = _equals ? _cache[0] : _cache[0] = props;
  return _equals ? _cache[1] : _cache[1] = _value.example;
}"
`;

exports[`expressions > should optimize new expressions 1`] = `
"import { useMemo as _useMemo } from \\"react\\";
import { $$cache as _$$cache } from \\"forgetti/runtime\\";
import { $$equals as _$$equals } from \\"forgetti/runtime\\";
function Example(props) {
  let _cache = _$$cache(_useMemo, 3),
    _value = 0 in _cache ? _cache[0] : _cache[0] = X,
    _equals = _$$equals(_cache, 1, props),
    _value2 = _equals ? _cache[1] : _cache[1] = props;
  return _equals ? _cache[2] : _cache[2] = new _value(_value2);
}"
`;

exports[`expressions > should optimize object expressions 1`] = `
"import { useMemo as _useMemo } from \\"react\\";
import { $$cache as _$$cache } from \\"forgetti/runtime\\";
import { $$equals as _$$equals } from \\"forgetti/runtime\\";
function Example(props) {
  let _cache = _$$cache(_useMemo, 8),
    _equals = _$$equals(_cache, 0, props),
    _value = _equals ? _cache[0] : _cache[0] = props,
    _value2 = _equals ? _cache[1] : _cache[1] = _value.a,
    _equals2 = _$$equals(_cache, 2, _value2),
    _value3 = _equals2 ? _cache[2] : _cache[2] = _value2,
    _value4 = _equals ? _cache[3] : _cache[3] = _value.b,
    _equals3 = _$$equals(_cache, 4, _value4),
    _value5 = _equals3 ? _cache[4] : _cache[4] = _value4,
    _value6 = _equals ? _cache[5] : _cache[5] = _value.c,
    _equals4 = _$$equals(_cache, 6, _value6),
    _value7 = _equals4 ? _cache[6] : _cache[6] = _value6;
  return _equals2 && _equals3 && _equals4 ? _cache[7] : _cache[7] = {
    a: _value3,
    b: _value5,
    ..._value7
  };
}"
`;

exports[`expressions > should optimize optional call expression 1`] = `
"import { useMemo as _useMemo } from \\"react\\";
import { $$cache as _$$cache } from \\"forgetti/runtime\\";
import { $$branch as _$$branch } from \\"forgetti/runtime\\";
import { $$equals as _$$equals } from \\"forgetti/runtime\\";
function Example(props) {
  let _cache = _$$cache(_useMemo, 7),
    _nullish,
    _equals = _$$equals(_cache, 0, props),
    _value = _equals ? _cache[0] : _cache[0] = props,
    _value2 = _equals ? _cache[1] : _cache[1] = _value.a,
    _hoisted = _nullish = _$$equals(_cache, 2, _value2) ? _cache[2] : _cache[2] = _value2,
    _equals3 = _$$equals(_cache, 3, _hoisted),
    _value5 = _equals3 ? _cache[3] : _cache[3] = _hoisted,
    _value4;
  if (_equals3 ? _cache[4] : _cache[4] = _value5 == null) {
    let _cache2 = _$$branch(_cache, 5, 1);
    _value4 = 0 in _cache2 ? _cache2[0] : _cache2[0] = void 0;
  } else {
    let _cache3 = _$$branch(_cache, 6, 7),
      _equals4 = _$$equals(_cache3, 0, _nullish),
      _value8 = _equals4 ? _cache3[0] : _cache3[0] = _nullish,
      _equals5 = _$$equals(_cache3, 1, props),
      _value9 = _equals5 ? _cache3[1] : _cache3[1] = props,
      _value10 = _equals5 ? _cache3[2] : _cache3[2] = _value9.b,
      _equals6 = _$$equals(_cache3, 3, _value10),
      _value11 = _equals6 ? _cache3[3] : _cache3[3] = _value10,
      _value12 = _equals5 ? _cache3[4] : _cache3[4] = _value9.c,
      _equals7 = _$$equals(_cache3, 5, _value12),
      _value13 = _equals7 ? _cache3[5] : _cache3[5] = _value12;
    _value4 = _equals4 && _equals && _equals6 && _equals7 ? _cache3[6] : _cache3[6] = _value8.call(_value, _value11, _value13);
  }
  return _value4;
}"
`;

exports[`expressions > should optimize optional member expression 1`] = `
"import { useMemo as _useMemo } from \\"react\\";
import { $$cache as _$$cache } from \\"forgetti/runtime\\";
import { $$branch as _$$branch } from \\"forgetti/runtime\\";
import { $$equals as _$$equals } from \\"forgetti/runtime\\";
function Example(props) {
  let _cache = _$$cache(_useMemo, 14),
    _nullish,
    _nullish2,
    _equals = _$$equals(_cache, 0, props),
    _value2 = _equals ? _cache[0] : _cache[0] = props,
    _value;
  if (_equals ? _cache[1] : _cache[1] = _value2 == null) {
    let _cache2 = _$$branch(_cache, 2, 1);
    _value = 0 in _cache2 ? _cache2[0] : _cache2[0] = void 0;
  } else {
    let _cache3 = _$$branch(_cache, 3, 2),
      _value5 = _$$equals(_cache3, 0, props) ? _cache3[0] : _cache3[0] = props;
    _value = _equals ? _cache3[1] : _cache3[1] = _value2.a;
  }
  let _hoisted2 = _nullish2 = _$$equals(_cache, 4, _value) ? _cache[4] : _cache[4] = _value,
    _equals4 = _$$equals(_cache, 5, _hoisted2),
    _value9 = _equals4 ? _cache[5] : _cache[5] = _hoisted2,
    _value8;
  if (_equals4 ? _cache[6] : _cache[6] = _value9 == null) {
    let _cache4 = _$$branch(_cache, 7, 1);
    _value8 = 0 in _cache4 ? _cache4[0] : _cache4[0] = void 0;
  } else {
    let _cache5 = _$$branch(_cache, 8, 2),
      _equals5 = _$$equals(_cache5, 0, _nullish2),
      _value12 = _equals5 ? _cache5[0] : _cache5[0] = _nullish2;
    _value8 = _equals5 ? _cache5[1] : _cache5[1] = _value12.b;
  }
  let _hoisted = _nullish = _$$equals(_cache, 9, _value8) ? _cache[9] : _cache[9] = _value8,
    _equals7 = _$$equals(_cache, 10, _hoisted),
    _value16 = _equals7 ? _cache[10] : _cache[10] = _hoisted,
    _value15;
  if (_equals7 ? _cache[11] : _cache[11] = _value16 == null) {
    let _cache6 = _$$branch(_cache, 12, 1);
    _value15 = 0 in _cache6 ? _cache6[0] : _cache6[0] = void 0;
  } else {
    let _cache7 = _$$branch(_cache, 13, 2),
      _equals8 = _$$equals(_cache7, 0, _nullish),
      _value19 = _equals8 ? _cache7[0] : _cache7[0] = _nullish;
    _value15 = _equals8 ? _cache7[1] : _cache7[1] = _value19.c;
  }
  return _value15;
}"
`;

exports[`expressions > should optimize sequence expressions 1`] = `
"import { useMemo as _useMemo } from \\"react\\";
import { $$cache as _$$cache } from \\"forgetti/runtime\\";
import { $$equals as _$$equals } from \\"forgetti/runtime\\";
function Example(props) {
  let _cache = _$$cache(_useMemo, 3),
    _equals = _$$equals(_cache, 0, props),
    _value = _equals ? _cache[0] : _cache[0] = props;
  return _equals ? _cache[1] : _cache[1] = _value.a(), _equals ? _cache[2] : _cache[2] = _value.b();
}"
`;

exports[`expressions > should optimize tagged templates 1`] = `
"import { useMemo as _useMemo } from \\"react\\";
import { $$cache as _$$cache } from \\"forgetti/runtime\\";
import { $$equals as _$$equals } from \\"forgetti/runtime\\";
function Example(props) {
  let _cache = _$$cache(_useMemo, 8),
    _equals = _$$equals(_cache, 0, props),
    _value = _equals ? _cache[0] : _cache[0] = props,
    _value2 = _equals ? _cache[1] : _cache[1] = _value.tag,
    _equals2 = _$$equals(_cache, 2, _value2),
    _value3 = _equals2 ? _cache[2] : _cache[2] = _value2,
    _value4 = _equals ? _cache[3] : _cache[3] = _value.a(),
    _equals3 = _$$equals(_cache, 4, _value4),
    _value5 = _equals3 ? _cache[4] : _cache[4] = _value4,
    _value6 = _equals ? _cache[5] : _cache[5] = _value.b(),
    _equals4 = _$$equals(_cache, 6, _value6),
    _value7 = _equals4 ? _cache[6] : _cache[6] = _value6;
  return _equals2 && _equals3 && _equals4 ? _cache[7] : _cache[7] = _value3\`\${_value5}, \${_value7}\`;
}"
`;

exports[`expressions > should optimize template literals 1`] = `
"import { useMemo as _useMemo } from \\"react\\";
import { $$cache as _$$cache } from \\"forgetti/runtime\\";
import { $$equals as _$$equals } from \\"forgetti/runtime\\";
function Example(props) {
  let _cache = _$$cache(_useMemo, 6),
    _equals = _$$equals(_cache, 0, props),
    _value = _equals ? _cache[0] : _cache[0] = props,
    _value2 = _equals ? _cache[1] : _cache[1] = _value.a(),
    _equals2 = _$$equals(_cache, 2, _value2),
    _value3 = _equals2 ? _cache[2] : _cache[2] = _value2,
    _value4 = _equals ? _cache[3] : _cache[3] = _value.b(),
    _equals3 = _$$equals(_cache, 4, _value4),
    _value5 = _equals3 ? _cache[4] : _cache[4] = _value4;
  return _equals2 && _equals3 ? _cache[5] : _cache[5] = \`\${_value3}, \${_value5}\`;
}"
`;

exports[`expressions > should optimize unary expressions 1`] = `
"import { useMemo as _useMemo } from \\"react\\";
import { $$cache as _$$cache } from \\"forgetti/runtime\\";
import { $$equals as _$$equals } from \\"forgetti/runtime\\";
function Example(props) {
  let _cache = _$$cache(_useMemo, 4),
    _equals = _$$equals(_cache, 0, props),
    _value = _equals ? _cache[0] : _cache[0] = props,
    _value2 = _equals ? _cache[1] : _cache[1] = _value.a,
    _equals2 = _$$equals(_cache, 2, _value2),
    _value3 = _equals2 ? _cache[2] : _cache[2] = _value2;
  return _equals2 ? _cache[3] : _cache[3] = !_value3;
}"
`;
